File: front/web/lib/stack-manifest.ts
/**
 * stack-manifest.ts — the PURE TS twin of the authoritative Go Kernel source
 * `back/kernel/stackmanifest` (DP02 — the StackManifest engraved as a
 * first-class Kernel SOURCE, ROADMAP-provisioning-deploy EPIC A).
 *
 * DETERMINISM-FIRST (CLAUDE.md §6/§8): validation, canonicalization and
 * content-addressing are pure functions — the GO CODE IS AUTHORITATIVE; this
 * twin is byte-parity-pinned to it by the vitest mirror (the seeded Example
 * manifest must hash to the EXACT hash the Go side measured —
 * GO_MANIFEST_HASH in stack-manifest.test.ts). Any one-byte divergence reds.
 *
 * Canonical form = records.Canonicalize (S02 reused, never forked): the JSON
 * body with object keys sorted lexicographically (recursively), no
 * insignificant whitespace, empty/omitted fields dropped (Go omitempty).
 * Content address = sha256 hex of those bytes (records.Hash).
 *
 * THE WALL (CLAUDE.md §2): stack_manifest is ABOVE-the-line truth. This module
 * never writes — the Workbench /stack-manifest route only READS the seeded
 * manifest and RE-MEASURES (validate + hash); engraving a manifest flows
 * through idea → mirror → /goal → human approval, never a screen.
 */

/** The CLOSED role set (DP02 — an unknown role is refused, never guessed). */
export const ROLES = [
	"server",
	"datastore",
	"cache",
	"pooler",
	"workflow",
	"bus",
	"observability",
	"errortracking",
	"git",
	"tickets",
	"auth",
	"docs",
	"connector",
	"interpreter",
] as const;
export type Role = (typeof ROLES)[number];

/** The CLOSED profile set (SPEC-stack-2026 §one-shot, compose à profils). */
export const PROFILES = [
	"core",
	"docs",
	"observability",
	"qa",
	"git",
	"tickets",
	"connectors",
	"non-prod",
	"full",
] as const;
export type Profile = (typeof PROFILES)[number];

export interface Service {
	name: string;
	role: string;
	/** container image — empty when the service is EMITTED (DP03+). */
	image?: string;
	internal_port: number;
	profile: string;
	healthcheck?: string;
	depends_on?: string[];
}

export interface Volume {
	name: string;
	/** ENV VAR REFERENCE for the bind device — never a hardcoded path. */
	device_var: string;
}

export interface Network {
	name: string;
	external: boolean;
}

/** The DP02 first-class Kernel source: the DECLARED stack topology. */
export interface StackManifest {
	app: string;
	services: Service[];
	volumes?: Volume[];
	network: Network;
	connector_scopes?: string[];
}

/** The closed refusal codes (DP02 done-criteria) — mirrors Go. */
export type RefusalCode =
	| "STACK_NAME_REQUIRED"
	| "STACK_HAS_NO_SERVER"
	| "UNKNOWN_SERVICE_ROLE"
	| "DUPLICATE_INTERNAL_PORT"
	| "UNKNOWN_PROFILE"
	| "SERVICE_NAME_REQUIRED"
	| "DUPLICATE_SERVICE_NAME";

export interface Refusal {
	code: RefusalCode;
	message: string;
}

export function isKnownRole(r: string): r is Role {
	return (ROLES as readonly string[]).includes(r);
}

export function isKnownProfile(p: string): p is Profile {
	return (PROFILES as readonly string[]).includes(p);
}

/**
 * validate — the PURE total validator, the exact twin of Go
 * stackmanifest.Validate: name required, services named + unique, roles and
 * profiles inside their closed sets, internal ports unique, ≥1 role=server.
 * Returns null when valid, a Refusal (code + actionable message) otherwise.
 */
export function validate(m: StackManifest): Refusal | null {
	if (!m.app) {
		return {
			code: "STACK_NAME_REQUIRED",
			message: "the manifest must declare an app name",
		};
	}
	const seenNames = new Set<string>();
	const seenPorts = new Map<number, string>();
	let hasServer = false;
	for (const s of m.services) {
		if (!s.name) {
			return {
				code: "SERVICE_NAME_REQUIRED",
				message: "every service must be named",
			};
		}
		if (seenNames.has(s.name)) {
			return {
				code: "DUPLICATE_SERVICE_NAME",
				message: `service "${s.name}" declared twice`,
			};
		}
		seenNames.add(s.name);
		if (!isKnownRole(s.role)) {
			return {
				code: "UNKNOWN_SERVICE_ROLE",
				message: `service "${s.name}": role "${s.role}" is outside the closed set [${ROLES.join(" ")}]`,
			};
		}
		if (!isKnownProfile(s.profile)) {
			return {
				code: "UNKNOWN_PROFILE",
				message: `service "${s.name}": profile "${s.profile}" is outside the closed set [${PROFILES.join(" ")}]`,
			};
		}
		const prior = seenPorts.get(s.internal_port);
		if (prior !== undefined) {
			return {
				code: "DUPLICATE_INTERNAL_PORT",
				message: `services "${prior}" and "${s.name}" both declare internal port ${s.internal_port}`,
			};
		}
		seenPorts.set(s.internal_port, s.name);
		if (s.role === "server") {
			hasServer = true;
		}
	}
	if (!hasServer) {
		return {
			code: "STACK_HAS_NO_SERVER",
			message: "the stack must declare at least one role=server service",
		};
	}
	return null;
}

/** drops Go-omitempty values: "", undefined, empty arrays. */
function compact(v: unknown): unknown {
	if (Array.isArray(v)) {
		return v.map(compact);
	}
	if (v !== null && typeof v === "object") {
		const out: Record<string, unknown> = {};
		for (const [k, val] of Object.entries(v as Record<string, unknown>)) {
			if (val === undefined || val === "") continue;
			if (Array.isArray(val) && val.length === 0) continue;
			out[k] = compact(val);
		}
		return out;
	}
	return v;
}

/** encodeCanonical — keys sorted recursively, no whitespace (records.Canonicalize twin). */
function encodeCanonical(v: unknown): string {
	if (Array.isArray(v)) {
		return `[${v.map(encodeCanonical).join(",")}]`;
	}
	if (v !== null && typeof v === "object") {
		const keys = Object.keys(v as Record<string, unknown>).sort();
		return `{${keys
			.map(
				(k) =>
					`${JSON.stringify(k)}:${encodeCanonical((v as Record<string, unknown>)[k])}`,
			)
			.join(",")}}`;
	}
	return JSON.stringify(v);
}

/**
 * canonicalBody — the canonical JSONB bytes of a VALID manifest: the AST plus
 * the kind discriminator "stack_manifest", keys sorted, omitempty applied.
 * Throws the Refusal of an invalid manifest (an invalid manifest never gets
 * an address — Go parity).
 */
export function canonicalBody(m: StackManifest): string {
	const refusal = validate(m);
	if (refusal) {
		throw new Error(`stackmanifest: ${refusal.code}: ${refusal.message}`);
	}
	return encodeCanonical(compact({ kind: "stack_manifest", ...m }));
}

async function sha256hex(s: string): Promise<string> {
	const { createHash } = await import("node:crypto");
	return createHash("sha256").update(s, "utf8").digest("hex");
}

/** hashManifest — records.Hash(records.Canonicalize(body)): the content address. */
export async function hashManifest(m: StackManifest): Promise<string> {
	return sha256hex(canonicalBody(m));
}

/** One content-addressed, append-only kernel.stack_manifest row (id == version == hash). */
export interface StackManifestRecord {
	id: string;
	body: string;
	version: string;
}

/** newRecord — the content-addressed record of a valid manifest (never written here). */
export async function newRecord(
	m: StackManifest,
): Promise<StackManifestRecord> {
	const body = canonicalBody(m);
	const h = await sha256hex(body);
	return { id: h, body, version: h };
}

/**
 * exampleManifest — the pinned reference manifest, the EXACT twin of Go
 * stackmanifest.Example(): the minimal /data/dockers-convention stack — one
 * reverse-proxied server, one datastore, the Go interpreter sidecar (profile
 * core, ADR 0040 D7), one named bind volume, the external traefik_default
 * network, one declared connector scope.
 */
export function exampleManifest(): StackManifest {
	return {
		app: "alphashop",
		services: [
			{
				name: "app",
				role: "server",
				image: "node:22-alpine",
				internal_port: 3000,
				profile: "core",
				healthcheck: "wget -q --spider http://localhost:3000/health",
				depends_on: ["postgres"],
			},
			{
				name: "postgres",
				role: "datastore",
				image: "postgres:17-alpine",
				internal_port: 5432,
				profile: "core",
				healthcheck: "pg_isready -U app",
			},
			{
				name: "interpreter",
				role: "interpreter",
				image: "",
				internal_port: 8973,
				profile: "core",
				healthcheck: "wget -q --spider http://localhost:8973/health",
			},
		],
		volumes: [{ name: "app_data", device_var: "APP_DATA_PATH" }],
		network: { name: "traefik_default", external: true },
		connector_scopes: ["postgres:read-only"],
	};
}

File: back/kernel/stackmanifest/stackmanifest.go
// Package stackmanifest engraves the StackManifest as a first-class Kernel
// SOURCE (DP02, ROADMAP-provisioning-deploy EPIC A — unlocked by the DP01
// measured GO, ADR 0064).
//
// A StackManifest is a DECLARED DSL AST: app name, services (role inside a
// CLOSED set, internal port, profile inside a CLOSED set, healthcheck,
// depends_on), named bind volumes (env-var device references — never a
// hardcoded path, SPEC-stack-2026: no hardcoded URL/secret), the external
// reverse-proxy network and the declared connector scopes. The manifest
// DECLARES topology; it resolves no URL and no secret (that is the
// per-environment projection, EPIC B / DP06+).
//
// Content addressing REUSES records.Hash(records.Canonicalize(body)) (S02,
// never forked): id == version == the SHA-256 of the canonical JSONB body —
// append-only, the head moves by a NEW version. Validation is a PURE total
// function (determinism-first, CLAUDE.md §6/§8): name required, at least one
// role=server service, unique internal ports per app, roles and profiles
// inside their closed sets — an unknown role is a ValidationError
// (UNKNOWN_SERVICE_ROLE), never guessed.
//
// THE WALL (CLAUDE.md §2): stack_manifest is ABOVE-the-line truth. This
// package is pure (no DB, no I/O); the kernel.stack_manifest table (migration
// kernel_stack_manifest_baseline.sql) gives the agent role SELECT only —
// writing a manifest flows through idea → mirror → /goal → human approval.
package stackmanifest

import (
	"encoding/json"
	"fmt"

	"github.com/steph-frtech/aidos/back/kernel/records"
)

// KindStackManifest is the record kind discriminator carried by every
// stack_manifest body (the kernel.stack_manifest table of the additive Atlas
// migration). The seven KRDCore kinds (records.Kinds) stay closed —
// stack_manifest is a kernel DSL-AST kind like entity/policy/operation, with
// its own table and its own pure validator (this package).
const KindStackManifest = "stack_manifest"

// Role is the function a service plays in the emitted stack. The set is
// CLOSED (DP02): an unknown role is refused, never guessed.
type Role string

const (
	RoleServer        Role = "server"
	RoleDatastore     Role = "datastore"
	RoleCache         Role = "cache"
	RolePooler        Role = "pooler"
	RoleWorkflow      Role = "workflow" // Windmill — NEVER Temporal (SPEC-stack-2026)
	RoleBus           Role = "bus"
	RoleObservability Role = "observability"
	RoleErrorTracking Role = "errortracking"
	RoleGit           Role = "git"
	RoleTickets       Role = "tickets"
	RoleAuth          Role = "auth"
	RoleDocs          Role = "docs"
	RoleConnector     Role = "connector"
	RoleInterpreter   Role = "interpreter" // the Go sidecar (ADR 0040 D7)
)

var roleOrder = []Role{
	RoleServer, RoleDatastore, RoleCache, RolePooler, RoleWorkflow, RoleBus,
	RoleObservability, RoleErrorTracking, RoleGit, RoleTickets, RoleAuth,
	RoleDocs, RoleConnector, RoleInterpreter,
}

// Roles returns the closed role set in canonical order (enumerable, never
// invented ad hoc — the Workbench and the validator share this single list).
func Roles() []Role {
	out := make([]Role, len(roleOrder))
	copy(out, roleOrder)
	return out
}

// IsKnownRole reports whether r is inside the closed DP02 role set.
func IsKnownRole(r Role) bool {
	for _, k := range roleOrder {
		if k == r {
			return true
		}
	}
	return false
}

// Profile is the compose profile a service belongs to. The set is CLOSED
// (SPEC-stack-2026 §one-shot: core · docs · observability · qa · git ·
// tickets · connectors · non-prod · full).
type Profile string

const (
	ProfileCore          Profile = "core"
	ProfileDocs          Profile = "docs"
	ProfileObservability Profile = "observability"
	ProfileQA            Profile = "qa"
	ProfileGit           Profile = "git"
	ProfileTickets       Profile = "tickets"
	ProfileConnectors    Profile = "connectors"
	ProfileNonProd       Profile = "non-prod"
	ProfileFull          Profile = "full"
)

var profileOrder = []Profile{
	ProfileCore, ProfileDocs, ProfileObservability, ProfileQA, ProfileGit,
	ProfileTickets, ProfileConnectors, ProfileNonProd, ProfileFull,
}

// Profiles returns the closed profile set in canonical order.
func Profiles() []Profile {
	out := make([]Profile, len(profileOrder))
	copy(out, profileOrder)
	return out
}

// IsKnownProfile reports whether p is inside the closed SPEC-stack-2026
// profile set.
func IsKnownProfile(p Profile) bool {
	for _, k := range profileOrder {
		if k == p {
			return true
		}
	}
	return false
}

// Service is one declared service of the stack.
type Service struct {
	Name string `json:"name"`
	Role Role   `json:"role"`
	// Image is the container image — or empty when the service is EMITTED
	// (built from the app's own phase, DP03+).
	Image        string   `json:"image,omitempty"`
	InternalPort int      `json:"internal_port"`
	Profile      Profile  `json:"profile"`
	Healthcheck  string   `json:"healthcheck,omitempty"`
	DependsOn    []string `json:"depends_on,omitempty"`
}

// Volume is a named bind volume per the /data/dockers convention. DeviceVar is
// the ENV VAR REFERENCE for the bind device (e.g. "APP_DATA_PATH") — never a
// hardcoded path.
type Volume struct {
	Name      string `json:"name"`
	DeviceVar string `json:"device_var"`
}

// Network is the reverse-proxy network (traefik_default — external everywhere
// except the traefik deployment that owns it).
type Network struct {
	Name     string `json:"name"`
	External bool   `json:"external"`
}

// StackManifest is the DP02 first-class Kernel source: the declared topology
// of an emitted stack. It declares; it never resolves (no URL, no secret —
// the per-environment projection does that, EPIC B).
type StackManifest struct {
	AppName         string    `json:"app"`
	Services        []Service `json:"services"`
	Volumes         []Volume  `json:"volumes,omitempty"`
	Network         Network   `json:"network"`
	ConnectorScopes []string  `json:"connector_scopes,omitempty"`
}

// ValidationError is the actionable refusal of an invalid manifest (the
// BlockReason discipline: a code, never a guess).
type ValidationError struct {
	Code    string
	Message string
}

func (e *ValidationError) Error() string {
	return fmt.Sprintf("stackmanifest: %s: %s", e.Code, e.Message)
}

// The closed refusal codes (DP02 done-criteria).
const (
	CodeStackNameRequired    = "STACK_NAME_REQUIRED"
	CodeStackHasNoServer     = "STACK_HAS_NO_SERVER"
	CodeUnknownServiceRole   = "UNKNOWN_SERVICE_ROLE"
	CodeDuplicateInternal    = "DUPLICATE_INTERNAL_PORT"
	CodeUnknownProfile       = "UNKNOWN_PROFILE"
	CodeServiceNameRequired  = "SERVICE_NAME_REQUIRED"
	CodeDuplicateServiceName = "DUPLICATE_SERVICE_NAME"
)

// Validate is the PURE total validator of a StackManifest (determinism-first):
// name required, every service named and unique, roles and profiles inside
// their closed sets, internal ports unique per app, at least one role=server.
// It never touches the DB and never writes truth.
func Validate(m StackManifest) error {
	if m.AppName == "" {
		return &ValidationError{
			Code:    CodeStackNameRequired,
			Message: "the manifest must declare an app name",
		}
	}
	seenNames := map[string]bool{}
	seenPorts := map[int]string{}
	hasServer := false
	for _, s := range m.Services {
		if s.Name == "" {
			return &ValidationError{
				Code:    CodeServiceNameRequired,
				Message: "every service must be named",
			}
		}
		if seenNames[s.Name] {
			return &ValidationError{
				Code:    CodeDuplicateServiceName,
				Message: fmt.Sprintf("service %q declared twice", s.Name),
			}
		}
		seenNames[s.Name] = true
		if !IsKnownRole(s.Role) {
			return &ValidationError{
				Code: CodeUnknownServiceRole,
				Message: fmt.Sprintf("service %q: role %q is outside the closed set %v",
					s.Name, s.Role, roleOrder),
			}
		}
		if !IsKnownProfile(s.Profile) {
			return &ValidationError{
				Code: CodeUnknownProfile,
				Message: fmt.Sprintf("service %q: profile %q is outside the closed set %v",
					s.Name, s.Profile, profileOrder),
			}
		}
		if prior, dup := seenPorts[s.InternalPort]; dup {
			return &ValidationError{
				Code: CodeDuplicateInternal,
				Message: fmt.Sprintf("services %q and %q both declare internal port %d",
					prior, s.Name, s.InternalPort),
			}
		}
		seenPorts[s.InternalPort] = s.Name
		if s.Role == RoleServer {
			hasServer = true
		}
	}
	if !hasServer {
		return &ValidationError{
			Code:    CodeStackHasNoServer,
			Message: "the stack must declare at least one role=server service",
		}
	}
	return nil
}

// body is the JSON record body of a manifest — the AST plus the kind
// discriminator, the shape stored in kernel.stack_manifest.
type body struct {
	Kind string `json:"kind"`
	StackManifest
}

// CanonicalBody returns the canonical JSONB bytes of a VALID manifest —
// records.Canonicalize over the kind-discriminated body (S02 reused, never
// forked). Same manifest → same bytes, always.
func CanonicalBody(m StackManifest) ([]byte, error) {
	if err := Validate(m); err != nil {
		return nil, err
	}
	raw, err := json.Marshal(body{Kind: KindStackManifest, StackManifest: m})
	if err != nil {
		return nil, err
	}
	return records.Canonicalize(raw)
}

// HashManifest returns the content address of a valid manifest:
// records.Hash(records.Canonicalize(body)).
func HashManifest(m StackManifest) (string, error) {
	canon, err := CanonicalBody(m)
	if err != nil {
		return "", err
	}
	return records.Hash(canon), nil
}

// Record is one content-addressed, append-only row of kernel.stack_manifest —
// the same five-column shape as every KRDCore record table (S02). id ==
// version == Hash(Canonicalize(body)); the head moves by inserting a NEW row.
type Record struct {
	ID           string          `json:"id"`
	Body         json.RawMessage `json:"body"`
	Version      string          `json:"version"`
	SupersededBy string          `json:"superseded_by,omitempty"`
}

// NewRecord builds the content-addressed record of a manifest. It validates
// first (an invalid manifest never gets an address), canonicalizes, and sets
// id == version == the content hash. It does NOT write to the DB — the wall:
// only the aidos CLI role writes kernel.stack_manifest, via an approved
// ChangeSet (idea → mirror → /goal → approval).
func NewRecord(m StackManifest) (Record, error) {
	canon, err := CanonicalBody(m)
	if err != nil {
		return Record{}, err
	}
	h := records.Hash(canon)
	return Record{ID: h, Body: canon, Version: h}, nil
}

// Example returns the pinned reference manifest the Workbench /stack-manifest
// route seeds and the e2e proves: the minimal /data/dockers-convention stack —
// one reverse-proxied server, one datastore, the Go interpreter sidecar
// (profile core, ADR 0040 D7), one named bind volume, the external
// traefik_default network, one declared connector scope. Deterministic — a
// fixture, never drawn.
func Example() StackManifest {
	return StackManifest{
		AppName: "alphashop",
		Services: []Service{
			{
				Name:         "app",
				Role:         RoleServer,
				Image:        "node:22-alpine",
				InternalPort: 3000,
				Profile:      ProfileCore,
				Healthcheck:  "wget -q --spider http://localhost:3000/health",
				DependsOn:    []string{"postgres"},
			},
			{
				Name:         "postgres",
				Role:         RoleDatastore,
				Image:        "postgres:17-alpine",
				InternalPort: 5432,
				Profile:      ProfileCore,
				Healthcheck:  "pg_isready -U app",
			},
			{
				Name:         "interpreter",
				Role:         RoleInterpreter,
				Image:        "",
				InternalPort: 8973,
				Profile:      ProfileCore,
				Healthcheck:  "wget -q --spider http://localhost:8973/health",
			},
		},
		Volumes: []Volume{
			{Name: "app_data", DeviceVar: "APP_DATA_PATH"},
		},
		Network:         Network{Name: "traefik_default", External: true},
		ConnectorScopes: []string{"postgres:read-only"},
	}
}
